README.md
# Prismo for VSCode

[![Visual Studio Marketplace](https://img.shields.io/vscode-marketplace/d/guywaldman.prismo.svg)](https://marketplace.visualstudio.com/items?itemName=guywaldman.prismo)
[![Travis branch](https://img.shields.io/travis/guywald1/vscode-prismo/master.svg)](https://travis-ci.org/guywald1/vscode-prismo)

Visual Studio Code extension to beautify your titles, sections and separators.

[See on VSCode Marketplace](https://marketplace.visualstudio.com/items?itemName=guywaldman.prismo)

[See on GitHub](https://github.com/guywald1/vscode-prismo)

![Preview](https://media.giphy.com/media/xT9IgC3hcjvU59RnfW/giphy.gif)

# Motivation
I like neat and clean source code. It helps to visually separate sections in files (which arguably would be better off in a separate file on occasion) and I don't like manually aligning my titles and conforming to my OCD.

> **Note:** full-width spanning titles can be obtrusive and are recommended to be used with caution. In some instances, in my opinion, they can be helpful.
> An option to configure the width on all title variations is available as `width` (0 for full width).

# Introduction
> You may be familiar with my previous plugins that achieved similar goals - [AutoSect for Atom](https://github.com/guywald1/auto-sect) and [Prismo for vim](https://github.com/guywald1/vim-prismo).

> For future reference, I will call a section/title/separator by the generic name of _title_.

Who among has not wished for an easy way to format our titles? Sometimes, we just want to annotate a certain section of our code in a beautiful manner.

This extension will help you format your source code titles as you see fit. It is highly configurable so you may set them according to your personal tastes.

There are three variations:

### Normal

This title is the regular title you would get when calling `> Prismo: Decorate Title`.
By default, it spans across the entire document and should be used for top-level separation.

### Light and Hair

These title variations are for subtitles.

IMO, _light_ should be used for subtitles and _hair_ for separating your inner-most comment regions.

# Configuration

You have several options to configure Prismo in your `settings.json`.
The top-level ones are for the _normal_ title (calling `> Prismo: Decorate Title`):

### `prismo.dash`

The character to be used as a dash.

**type**: _string_

**Default**: _-_

### `prismo.width`

How wide the title should span.

If it is 0, the title will span across the entire editor width.

**type**: _number_

**Default**: 0

### `prismo.padding`

Padding between the title and the dashes.

> i.e. with a padding of _4_:
> ```
> /* ---    myTitle    ---
> ```

**type**: _number_

**Default**: 1

### `prismo.shouldUppercase`

Whether the title should be uppercased or not.

**type**: _boolean_

**Default**: true

## `prismo.light` and `prismo.hair`

These are configurations for your title variatons.
Both are objects containing the properties:

- dash
- padding
- shouldUppercase
- width

All are identical to the top-level ones, and are specific to each title variation.

However, they have different defaults.

### `prismo.light`

**Default (overrides)**:

**width**: 40
**shouldUppercase**: false

### `prismo.hair`

**Default (overrides)**:

**width**: 30
**shouldUppercase**: false

# Future Improvements

## Region Sectioning

_VSCode_, among other code editors, allows the use of [region](https://code.visualstudio.com/docs/editor/codebasics#_folding) for folding.
The user will be able to execute a command for selecting multiple lines, entering the desired title, and adding comments for making it a region.

  > i.e. seleting:
  > ```javascript
  > const obj = { foo: 'bar' }
  > console.log(obj.foo)
  > ```
  > and entering _myTitle_ would replace the selected lines into something like:
  > ```javascript
  > // ------- myTitle -------
  > 
  > //#region myTitle
  > const obj = { foo: 'bar' }
  > console.log(obj.foo)
  > //#endregion myTitle
  > ```

## Colored Titles

The user will be able to configure colors for their titles (similarly to the [todo-highlight](https://github.com/wayou/vscode-todo-highlight) extension).

## Multiple Variations

The user will be able to set more than three variations for titles and subtitles.



CHANGELOG.md
# Change Log
All notable changes to the "prismo" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]
- Initial release

src/utils/config.ts
import { window, workspace, WorkspaceConfiguration } from 'vscode'

// TODO: document
export interface TitleConfig {
  padding: number
  dash: string
  shouldUppercase: boolean,
  width: number
}


// TODO: document
export interface Config extends TitleConfig {
  light: TitleConfig
  hair: TitleConfig
}

// TODO: document
export enum Level {
  normal,
  light,
  hair
}

export const configOptions: string[] = [
  'padding',
  'dash',
  'shouldUppercase',
  'width',
  'light',
  'hair'
]

const defaultConfigTopLevel: TitleConfig = {
  padding: 1,
  dash: '-',
  shouldUppercase: true,
  width: 0
}
const defaultConfigSubtitle: TitleConfig = defaultConfigTopLevel

export const defaultConfig: Config = Object.assign({}, defaultConfigTopLevel, {
  light: Object.assign({}, defaultConfigSubtitle, {
    width: 40,
    shouldUppercase: false
  }),
  hair: Object.assign({}, defaultConfigSubtitle, {
    width: 30,
    shouldUppercase: false
  })
})

/**
 * Returns configuration object
 * Note:  This is unmemoized and is expensive (configuration includes nested objects).
 *        Therefore, call with caution.
 * @export
 * @returns {object} configuration object
 */
export function getConfig(): Config {
  const config: WorkspaceConfiguration = workspace.getConfiguration(
    'prismo',
    window.activeTextEditor.document.uri
  )
  return <Config>configOptions.reduce((acc: object, key: string) => {
    const value = config.get(key) || defaultConfig[key]
    if (value !== undefined) {
      return Object.assign(acc, { [key]: value })
    }
    return acc
  }, {})
}
